modules/create_timepoint/php/timepoint.class.inc

```diff
@@ -0,0 +1,344 @@
+        return (
+            $user->hasPermission('data_entry') &&
+            $candidate->isAccessibleBy($user)
+        );
+    }
+
+    /**
+     * Validate the post data,
+     * from the create timepoint form.
+     *
+     * @param array $values the form values.
+     *
+     * @return array
+     */
+    private function _validate(array $values) : array
+    {
+        $user = \User::singleton();
+
+        if (!isset($values['identifier'])) {
+            $values['error'] = 'BadRequest';
+            return $values;
+        }
+        $values['identifier'] = isset($values['identifier']) ?
+            intval($values['identifier']) : null;
+        $values['psc']        = empty($values['psc']) ?
+            null : $values['psc'];
+
+        // validate site entered
+        $site = (int) $values['psc'];
+        $user_list_of_sites = $user->getData('CenterIDs');
+        $num_sites          = count($user_list_of_sites);
+        $conflict           = [];
+        if ($num_sites == 0) {
+            $conflict['psc']
+                = "You must be affiliated with a site to create a timepoint.";
+        } else if ($num_sites > 1 && (empty($site) || !$user->hasCenter($site))) {
+            $conflict['psc']
+                = "Site must be selected from the available dropdown.";
+        }
+
+        //Project
+        $user_list_of_projects = $user->getProjectIDs();
+        if (count($user_list_of_projects) == 0) {
+            $conflict['project']
+                = "You must be affiliated with a project to create a timepoint.";
+        } else if (count($user_list_of_projects) > 1
+            && is_null($values['project'])
+        ) {
+            $conflict['project']
+                = "A project must be selected from the available dropdown.";
+        }
+
+        $candid       = $values['candID'];
+        $subprojectID = intval($values['subproject']);
+        $visitLabel   = $values['visit'] ?? '';
+
+        try {
+            \TimePoint::isValidVisitLabel(
+                new CandID($candid),
+                $subprojectID,
+                $visitLabel
+            );
+        } catch (\LorisException $exception) {
+            $conflict['visitLabel'] = $exception->getMessage();
+        }
+
+        // validate language
+        $languages = \Utility::getLanguageList();
+        if (count($languages) > 1 && empty($values['languages'])) {
+            $conflict['languages'] = 'A language must be selected.';
+        }
+        // set form message if exists.
+        if ($conflict) {
+            $values['conflict'] = $conflict;
+        }
+        return $values;
+    }
+
+    /**
+     * Returns true if the user has permission to access
+     * the Create Timepoint module
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool true if user has permission
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return $user->hasPermission('data_entry');
```

Comment:
should this maybe also have `$this->candidate->isAccessibleBy($user)`?

Reply:
@zaliqarosli I asked @xlecours (i think last week) about how to access `candidate` in this function because the identifier needed to make the `candidate` isn't accessible until the `handle` function gets called after the _hasAccess function. I believe he said the _hasAccess function should just be checking if the module is accessible for the user and not checking whether the user can carry out what the requests are meant for doing.

Reply:
okay! i think this is fine. i'll do a final test!